### ISSUE TITLE
[clang-tidy] modernize-use-std-print misbehaves when arguments must be re-ordered and casts added or c_str() removed

### ISSUE THREAD

**Issue by mikecrowe** [llvm/llvm-project#64033]
The newly-added _modernize-use-std-print_ check gets confused if it has to re-order arguments at the same time as adding casts or removing calls to `c_str()`. For example applying the check with _StrictMode_ enabled to:
```c++
printf("%*s=%*d\n", 4, s.c_str(), 3, ui);
```
yields:
```c++
std::println("{:>{}}={:{}}", s.c_str(), s4, ui, static_cast<int>(3));
```
rather than the expected:
```c++
std::println("{:>{}}={:{}}", s, 4, static_cast<int>(ui), 3);
```
I attempted to fix this by:

1. Doing the argument re-ordering first in `FormatStringConverter::applyFixes`.
2. Storing the argument index in `FormatStringConverter::ArgFixes` rather than the `Expr *` for the argument, and changing the index as required during re-ordering.

This meant that the cast was being applied to the correct argument, but since there was already a replacement pending for that argument functions like `getBeginLoc()` don't report the right position and the cast ends up in the wrong place resulting in:
```c++
std::println("{:>{}}={:{}}", s.c_str(), 4s, uistatic_cast<int>(), 3);
```
Note that I hadn't even attempted to fix the `c_str()` removal at this point. I've no idea where to even start with that since `FormatStringConverter::ArgCstrRemovals` contains the matched nodes rather than an index.

Is there a good way to have fixes built upon earlier fixes within the same `DiagnosticBuilder` instance, or do I need to handle all the possible combinations myself and generate a single combined fix for each?

An alternative would be to leave the arguments where they are and use argument indices in the format string, e.g.:
```c++
std::println("{1:>{0}}={3:{2}}", 4, s, 3, static_cast<int>(ui));
```
However, if argument indices are used then they must be used in the entire format string and I wouldn't necessarily know whether they needed to be used until I'd already started rewriting earlier arguments which makes even this rather messy.

Any advice gratefully received.

**Comment by mikecrowe** [comment]
>  However, if argument indices are used then they must be used in the entire format satring and I wouldn't necessarily know whether they needed to be used until I'd already started rewriting earlier arguments which makes even this rather messy.

It turns out not to be as messy to implement as I expected provided the argument indices are added whenever argument re-ordering would be necessary, even if no `c_str()`-removal or casts need to be added. However, it does mean that the results of the check are a bit more ugly and confusing.

**Comment by mikecrowe** [comment]
Potential fix in [D156616](https://reviews.llvm.org/D156616).
